products/BellHybrid/apps/common/src/widgets/ListViewWithLabels.cpp

```diff
@@ -18,125 +23,133 @@ namespace gui
         body->dimensionChangedCallback = [&](gui::Item &, const BoundingBox &newDim) -> bool { return true; };
     }
 
-    void ListViewWithLabels::addItemsOnPage()
+    void ListViewWithLabels::addLabelItem()
     {
-        currentPageSize = 0;
-        itemsOnPage     = 0;
-        labelAdded      = false;
+        if (!labelsCount) {
+            return;
+        }
 
-        ListItem *item;
-        while ((item = provider->getItem(getOrderFromDirection())) != nullptr) {
-            /* If direction is top-to-bottom, add label mark before adding relaxation item. */
-            if (direction == listview::Direction::Bottom) {
-                addLabelMarker(item);
-            }
-            /* Check if new item fits, if it does - add it, if not - handle label insertion
-             * case for bottom-to-top navigation direction. */
-            if (getSlotsLeft() > 0) {
-                body->addWidget(item);
-                itemsOnPage++;
+        std::int32_t size = currentPageSize;
+        size              = direction == listview::Direction::Top ? size * -1 : size;
+
+        auto position = 0U;
+        for (std::uint32_t i = 0; i < labelsCount; ++i) {
+            // First label always starts from 0
+            if (i == 0) {
+                position = 0;
             }
             else {
-                /* Add invisible item to list to avoid memory leak */
-                item->setVisible(false);
-                body->addWidget(item);
-                break;
+                position += labelFiles[i - 1].second;
             }
-            /* If direction is bottom-to-top, add label mark after adding relaxation item. */
-            if (direction == listview::Direction::Top) {
-                addLabelMarker(item);
+            const auto [labelName, filesCount] = labelFiles[i];
+            if (((startIndex + size) == position) && filesCount && !labelAdded) {
```

Comment:
```suggestion
            if (((startIndex + size) == position) && (filesCount > 0) && !labelAdded) {
```

That way it's easier to read and to know that `filesCount` is not `std::optional`, pointer or some other item with overloaded `bool` operator :wink:

Reply:
You're right but `filesCount` is above so it's clear what is going on. Anyway I changed as you suggested ;)